parachains/runtimes/assets/statemine/tests/tests.rs

```diff
@@ -302,3 +302,70 @@ fn test_that_buying_ed_refund_does_not_refund() {
 			assert_eq!(Assets::total_supply(1), ExistentialDeposit::get());
 		});
 }
+
+#[test]
+fn test_asset_xcm_trader_not_possible_for_non_sufficient_assets() {
+	ExtBuilder::<Runtime>::default()
+		.with_collators(vec![AccountId::from(ALICE)])
+		.with_session_keys(vec![(
+			AccountId::from(ALICE),
+			AccountId::from(ALICE),
+			SessionKeys { aura: AuraId::from(sp_core::sr25519::Public::from_raw(ALICE)) },
+		)])
+		.build()
+		.execute_with(|| {
+			// Create a non-sufficient asset
+			// We set existential deposit to be identical to the one for Balances first
+			assert_ok!(Assets::force_create(
+				RuntimeHelper::<Runtime>::root_origin(),
+				1,
+				AccountId::from(ALICE).into(),
+				false,
+				ExistentialDeposit::get()
+			));
+
+			// We first mint enough asset for the account to exist for assets
+			assert_ok!(Assets::mint(
+				RuntimeHelper::<Runtime>::origin_of(AccountId::from(ALICE)),
+				1,
+				AccountId::from(ALICE).into(),
+				ExistentialDeposit::get()
+			));
+
+			let mut trader = <XcmConfig as xcm_executor::Config>::Trader::new();
+
+			// Set Alice as block author, who will receive fees
+			RuntimeHelper::<Runtime>::run_to_block(2, Some(AccountId::from(ALICE)));
+
+			// We are going to buy 4e9 weight
+			let bought = 4_000_000_000u64;
+
+			// lets calculate amount needed
+			let amount_needed = WeightToFee::weight_to_fee(&Weight::from_ref_time(bought));
```

Comment:
@joepetrowski 
good point, I guess, looks like an another confusion in all of those tests,

`WeightToFee` works with runtime's `Balance`, right?

so, the `amount_needed` is meant to be `DOT or KSM`,
```
let amount_needed = WeightToFee::weight_to_fee(&Weight::from_ref_time(bought));
```

but the value of `amount_needed` is used as a amount of asset:
```
let asset: MultiAsset = (asset_multilocation, amount_needed).into();
```

maybe correct/exact would be something like this:
```
type Dex = AssetFeeAsExistentialDepositMultiplier< 
        Runtime,
        WeightToFee,
        pallet_assets::BalanceToAssetBalance<Balances, Runtime, ConvertInto>,
>;

let local_asset_id = 1,;
let amount_needed = Dex::charge_weight_in_fungibles(local_asset_id, Weight::from_ref_time(bought));
```

do we want to change it here for all tests?

Reply:
Yeah, that makes more sense. I wouldn't call it `type Dex` though, as it implies an asset exchange could actually take place. It's really declaring, "how should I convert an amount in DOT to an amount in the asset".

The complicated part is that, if we had a synchronous DEX, we actually could convert the asset using `swap_asset_for_exact(asset: AssetId, fee_amount: Balance)`, but we're not there yet. Of course, actually making the swap would itself consume more weight, so there could be an option to fetch the spot rate on a DEX but still just convert the fee amount into some amount of fungible asset, and take that as a fee.

We should consider `ChargeWeightInFungibles` more for another PR, but I think for where we are right now (in that we are converting DOT fee to asset balance, but not actually providing the ability to swap it), we should look at this as a `Convert` type of item. `AssetFeeAsExistentialDepositMultiplier` specifies the means (just take the ratio of min balance / existential deposit).

Perhaps, `type FeeItemConversion` ("item" implying that ownership of a non-fungible could even be a "ticket" so some execution rights)? But before renaming some of the existing Config interface I'd want to dive a bit deeper on the issue and where we see it going.

Reply:
If I recall correctly, the tests did set the same ED for the asset and DOT/KSM, hence the reason we can use `
` as is.

Reply:
yes, I see now, all the tests works ok, because if we create asset with `ExistentialDeposit::get()`,
then the `BalanceToAssetBalance` return the same value, because of multiplier is `1` `// balance * asset.min_balance / min_balance`

so these two return the same value
```
let amount_needed = WeightToFee::weight_to_fee(&Weight::from_ref_time(bought));
let amount_needed2 = AssetFeeAsExistentialDepositMultiplierFeeCharger::charge_weight_in_fungibles(local_asset_id, Weight::from_ref_time(bought)).expect("aaa");
```

Reply:
@joepetrowski 
I changed `WeightToFee` with `AssetFeeAsExistentialDepositMultiplierFeeCharger` + some cleaning in all `test_asset_xcm_trader` tests, 

so, I am done here with this PR (if there are no more comments), not to block release or anything, 
and at least, now I understand all this stuff and I will later look at other tests in following issue: https://github.com/paritytech/cumulus/issues/1915 , ok?

Reply:
Yes, sounds good